### ISSUE TITLE
Investigate whether usage of CDI can be avoided

### ISSUE THREAD

**Issue by gunnarmorling** [mincong-h/gsoc-hsearch#86]
See https://github.com/mincong-h/gsoc-hsearch/blob/master/core/src/main/java/org/hibernate/search/jsr352/MassIndexerImpl.java#L236-L264

**Comment by gunnarmorling** [comment]
See https://github.com/mincong-h/gsoc-hsearch/blob/master/core/src/main/java/org/hibernate/search/jsr352/MassIndexerImpl.java#L236-L264

**Comment by mincong-h** [comment]
I've create a branch for this issue, called `issue-86`. The main pb is how to handle CDI in Java SE environment.

**Comment by mincong-h** [comment]
done.

**Comment by gunnarmorling** [comment]
Would still be nice to find out whether and how CDI could be made optional.

**Comment by yrodiere** [comment]
There are two things we need to distinguish:

 1. The dependency on `javax.inject`
 2. The reliance on CDI to resolve bean references inside the batch-job XML file

The first one cannot be avoided since JSR-352 relies on `javax.inject` (to define injected batch properties in particular).
The second one can be avoided, I think.

So, according to JSR-352 ("10.5 Batch Artifact Loading"):

 * Users can choose how to resolve bean references: either they add a `META-INF/batch.xml` file to map reference to the fully qualified class name (does *anyone* want to do that?) or they use a custom loader (e.g. CDI or Spring DI)
 * But more importantly, when the JSR-352 implementation fails to find a bean for the given reference using the two mechanisms above, it must interpret the reference as a fully-qualified class name!

I think we should:

 * replace bean references in the batch job XML file by fully-qualified class names
 * remove all @Named annotations on our bean implementations
 * see if it works

If it works, I guess we're good to remove our explicit dependency on `javax.inject` and consider we don't rely on any other injection than the one provided by JSR-352 implementations for batch properties anymore.

**Comment by mincong-h** [comment]
I replaced the bean references in the batch job XML file by full-qualified class names and removed all the `@Named` annotations in commit 638eeae. However, the `EntityManagerFactory` cannot be injected anymore under WildFly:

```
11:58:31,122 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-2) WFLYSRV0027: Starting deployment of "RestartIT.war" (runtime-name: "RestartIT.war")
11:58:31,751 INFO  [org.jboss.as.jpa] (MSC service thread 1-8) WFLYJPA0002: Read persistence.xml for h2
11:58:31,938 INFO  [org.jboss.as.jpa] (ServerService Thread Pool -- 67) WFLYJPA0010: Starting Persistence Unit (phase 1 of 2) Service 'RestartIT.war#h2'
11:58:31,969 INFO  [org.hibernate.jpa.internal.util.LogHelper] (ServerService Thread Pool -- 67) HHH000204: Processing PersistenceUnitInfo [
	name: h2
	...]
11:58:32,047 INFO  [org.jboss.weld.deployer] (MSC service thread 1-7) WFLYWELD0003: Processing weld deployment RestartIT.war
11:58:32,182 INFO  [org.hibernate.Version] (ServerService Thread Pool -- 67) HHH000412: Hibernate Core {5.0.7.Final}
11:58:32,185 INFO  [org.hibernate.cfg.Environment] (ServerService Thread Pool -- 67) HHH000206: hibernate.properties not found
11:58:32,188 INFO  [org.hibernate.cfg.Environment] (ServerService Thread Pool -- 67) HHH000021: Bytecode provider name : javassist
11:58:32,275 INFO  [org.hibernate.orm.deprecation] (ServerService Thread Pool -- 67) HHH90000001: Found usage of deprecated setting for specifying Scanner [hibernate.ejb.resource_scanner]; use [hibernate.archive.scanner] instead
11:58:32,288 INFO  [org.hibernate.annotations.common.Version] (ServerService Thread Pool -- 67) HCANN000001: Hibernate Commons Annotations {5.0.1.Final}
11:58:32,501 INFO  [org.hibernate.validator.internal.util.Version] (MSC service thread 1-7) HV000001: Hibernate Validator 5.2.3.Final
11:58:32,655 INFO  [org.jboss.as.ejb3.deployment] (MSC service thread 1-7) WFLYEJB0473: JNDI bindings for session bean named 'MessageManager' in deployment unit 'deployment "RestartIT.war"' are as follows:

	java:global/RestartIT/MessageManager!org.hibernate.search.jsr352.test.MessageManager
	java:app/RestartIT/MessageManager!org.hibernate.search.jsr352.test.MessageManager
	java:module/MessageManager!org.hibernate.search.jsr352.test.MessageManager
	java:global/RestartIT/MessageManager
	java:app/RestartIT/MessageManager
	java:module/MessageManager

11:58:32,923 INFO  [org.infinispan.configuration.cache.EvictionConfigurationBuilder] (ServerService Thread Pool -- 67) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
11:58:32,925 INFO  [org.infinispan.configuration.cache.EvictionConfigurationBuilder] (ServerService Thread Pool -- 67) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
11:58:32,927 INFO  [org.infinispan.configuration.cache.EvictionConfigurationBuilder] (ServerService Thread Pool -- 67) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
11:58:32,927 INFO  [org.infinispan.configuration.cache.EvictionConfigurationBuilder] (ServerService Thread Pool -- 67) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
11:58:32,938 INFO  [org.jboss.weld.deployer] (MSC service thread 1-4) WFLYWELD0006: Starting Services for CDI deployment: RestartIT.war
11:58:33,044 INFO  [org.jboss.weld.Version] (MSC service thread 1-4) WELD-000900: 2.3.2 (Final)
11:58:33,110 INFO  [org.jboss.weld.deployer] (MSC service thread 1-1) WFLYWELD0009: Starting weld service for deployment RestartIT.war
11:58:33,382 INFO  [org.jboss.as.jpa] (ServerService Thread Pool -- 25) WFLYJPA0010: Starting Persistence Unit (phase 2 of 2) Service 'RestartIT.war#h2'
11:58:33,505 INFO  [org.jboss.as.clustering.infinispan] (ServerService Thread Pool -- 67) WFLYCLINF0002: Started client-mappings cache from ejb container
11:58:33,710 INFO  [org.hibernate.dialect.Dialect] (ServerService Thread Pool -- 25) HHH000400: Using dialect: org.hibernate.dialect.H2Dialect
11:58:33,719 WARN  [org.hibernate.dialect.H2Dialect] (ServerService Thread Pool -- 25) HHH000431: Unable to determine H2 database version, certain features may not work
11:58:33,785 INFO  [org.hibernate.envers.boot.internal.EnversServiceImpl] (ServerService Thread Pool -- 25) Envers integration enabled? : true
11:58:34,251 INFO  [org.hibernate.search.engine.Version] (ServerService Thread Pool -- 25) HSEARCH000034: Hibernate Search 5.5.1.Final
11:58:34,391 INFO  [org.hibernate.tool.hbm2ddl.SchemaExport] (ServerService Thread Pool -- 25) HHH000227: Running hbm2ddl schema export
11:58:34,398 INFO  [org.hibernate.tool.hbm2ddl.SchemaExport] (ServerService Thread Pool -- 25) HHH000230: Schema export complete
11:58:36,645 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 25) WFLYUT0021: Registered web context: /RestartIT
11:58:36,741 INFO  [org.jboss.as.server] (management-handler-thread - 4) WFLYSRV0010: Deployed "RestartIT.war" (runtime-name : "RestartIT.war")
11:58:39,778 INFO  [org.hibernate.search.jsr352.RestartIT] (default task-1) Job execution (id=1) has status STARTING. Thread sleeps 1000 ms...
11:58:39,785 ERROR [org.hibernate.search.jsr352.internal.JobContextSetupListener] (Batch Thread - 1) java.lang.NullPointerException
11:58:39,789 ERROR [org.jberet] (Batch Thread - 1) JBERET000007: Failed to run job BatchIndexingJob, , org.jberet.job.model.Job@c1c2d853: java.lang.NullPointerException
	at org.hibernate.search.jsr352.internal.JobContextSetupListener.beforeJob(JobContextSetupListener.java:65)
	at org.jberet.runtime.runner.JobExecutionRunner.run(JobExecutionRunner.java:56)
	at org.wildfly.extension.batch.jberet.impl.BatchEnvironmentService$WildFlyBatchEnvironment$1.run(BatchEnvironmentService.java:243)
	at org.wildfly.extension.requestcontroller.RequestController$QueuedTask$1.run(RequestController.java:497)
	at org.jberet.spi.JobExecutor$3.run(JobExecutor.java:161)
	at org.jberet.spi.JobExecutor$1.run(JobExecutor.java:99)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at org.jboss.threads.JBossThread.run(JBossThread.java:320)
```

This is very weird because the EMF does exist according to the log.

```
11:58:31,969 INFO  [org.hibernate.jpa.internal.util.LogHelper] (ServerService Thread Pool -- 67) HHH000204: Processing PersistenceUnitInfo [
	name: h2
	...]
```

And here's the excerpt of the integration test `RestartId.java`:

```java
@RunWith(Arquillian.class)
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class RestartIT {

	...

	@PersistenceUnit(unitName = "h2")
	private EntityManagerFactory emf;

	@Deployment
	public static WebArchive createDeployment() {
		WebArchive war = ShrinkWrap
				.create( WebArchive.class, RestartIT.class.getSimpleName() + ".war" )
				.addAsResource( "META-INF/persistence.xml" )
				.addAsResource( "META-INF/batch-jobs/make-deployment-as-batch-app.xml" ) // WFLY-7000
				.addAsWebInfResource( "jboss-deployment-structure.xml" )
				.addAsWebInfResource( EmptyAsset.INSTANCE, "beans.xml" )
				.addPackage( Message.class.getPackage() );
		return war;
	}

	@Test
	public void testJob() throws InterruptedException, IOException, ParseException {

		insertData();

		assertEquals( 0, messageManager.findMessagesFor( SDF.parse( "31/08/2016" ) ).size() );
		assertEquals( 0, messageManager.findMessagesFor( SDF.parse( "01/09/2016" ) ).size() );

		// The 1st execution. Keep it alive and wait Byteman to stop it
		long execId1 = BatchIndexingJob.forEntity( Message.class ).start();
		JobExecution jobExec1 = BatchRuntime.getJobOperator().getJobExecution( execId1 );
		jobExec1 = keepTestAlive( jobExec1 );

		// Restart the job. This is the 2nd execution.
		long execId2 = BatchIndexingJob.restart( execId1 );
		JobExecution jobExec2 = BatchRuntime.getJobOperator().getJobExecution( execId2 );
		jobExec2 = keepTestAlive( jobExec2 );
		...
	}
	...
}
```

**Comment by mincong-h** [comment]
I cannot find out why we have this NPE. I though about several possible reasons:

1. The EMF doesn't exist, so NPE. Obviously, this is not true, because the EMF is created and found according to log: 

        11:58:31,969 INFO  [org.hibernate.jpa.internal.util.LogHelper] (ServerService Thread Pool -- 67) HHH000204: Processing PersistenceUnitInfo [
                    name: h2
                    ...]

2. The empty beans XML, deployed during the integration test, has a different discovery mode than what we need. But this hypothesis is not true neither, because if the empty beans XML caused problem, the class `org.hibernate.search.jsr352.internal.JobContextSetupListener` won't even be found.

    ```java
    @Deployment
    public static WebArchive createDeployment() {
        WebArchive war = ShrinkWrap
                ...
                .addAsWebInfResource( EmptyAsset.INSTANCE, "beans.xml" )
    ```

Can anyone give me some suggestions?

**Comment by yrodiere** [comment]
My best guess is that JBeret is designed in such a way that injection of anything other than `@BatchParameter`s is only done when the beans are instantiated with the dependency injection framework (CDI, Spring DI, ...) in the first place, which is not the case when using fully-qualified class names. In other words, either the DI frameworks handles everything (instantiation + injection) or nothing.

Anyway... If we want to solve #163, I doubt we'll be able to use injection anymore, because we can't include the name of the persistence unit in an annotation. Maybe we should address #163 first, and get back to this ticket when we're done?

**Comment by yrodiere** [comment]
Fixed as part of #186
